### ISSUE TITLE
Further time_index selection improvement

### ISSUE THREAD

**Issue by heikkilv** [smartsdk/ngsi-timeseries-api#172]
The version 0.6 included update to the time_index selection so that you can specify a custom attribute that can be used for the time_index value. However, it would be useful if the given custom value would also be used to search in the attribute metadata. For example, if I have entity with an attribute
<pre>
"someAttribute": {
    "type": "Number",
    "value": 1.2
    "metadata": {
        "myTime": {
            "type": "DateTime",
            "value": "2019-03-01T12:00:00Z"
        }
    }
}
</pre>
I could specify the subscription header Fiware-TimeIndex-Attribute with the value myTime and the time_index would be set with the value "2019-03-01T12:00:00Z". I could create a separate attribute myTime, but this approach becomes unnecessarely complicated, if there are more than one attribute that have attribute specific timestamp.

Also, while the attribute TimeIndex is now automatically used for the time_index selection, the timestamp attribute, which the is used in the data models <https://fiware-datamodels.readthedocs.io/en/latest/index.html>, is not. timestamp should be automatically used either before TimeIndex or after TimeIndex.

**Comment by c0c0n3** [comment]
Excellent suggestions. Thank you so much @heikkilv! We'll definitely consider this and it's likely going to be implemented in a future release, but for the time being we have very little bandwidth available to work on improvements/new features. As always, pull requests are welcome :-)

**Comment by taliaga** [comment]
Hi, thanks for getting in touch!

For the first part, before discussing the implementation of such specification along with the `Fiware-TimeIndex-Attribute`, let me understand your use-case. The "metadata" option of the policy should be enough to pick those values you need, isn't it? QL would typically pick the latest of the timestamp "value" found in attributes' metadata (I assume you need the latest one). Do you have more than one maybe?

For the second part, which models are using `timestamp` as you describe? Is that the name of the attribute to look for?

**Comment by heikkilv** [comment]
As for the <code>timestamp</code> meta attribute, pretty much all of the data models use it. Including for example [WeatherObserved](https://fiware-datamodels.readthedocs.io/en/latest/Weather/WeatherObserved/doc/spec/index.html). Also, the [guidelines](https://fiware-datamodels.readthedocs.io/en/latest/guidelines/index.html) instructs the use of <code>timestamp</code> meta attribute for attribute specific measurement times.

Using the latest timestamp "value" found in the attributes metadata is what I was thinking with the <code>Fiware-TimeIndex-Attribute</code>. Similarly as the <code>TimeInstant</code> and <code>dateModified</code> are currently treated.

I have actually already modified the code for my own use. I can make a pull request about this in a couple of days.

**Comment by taliaga** [comment]
Thanks for the pointers. I didn't recall seeing those in the past.

Note if you remove the dateModified from your subscription and TimeInstant is not there in your notification, the following policy step should work for your use-case. Too complex, I know. I agree it's time to adjust and flexibilize even more the TimeIndex policy. There's time information coming from everywhere these days.

Thank you! PRs will be more than welcome to open the discussion

**Comment by taliaga** [comment]
@heikkilv thanks for the PR!. Please see my latest comments, specially https://github.com/smartsdk/ngsi-timeseries-api/pull/176#discussion_r268629515 before we can merge it.

A departing question, how did you manage to get an attribute embedded in the metadata section of another attribute? What do you have (orion, agent) and how did you configure it to end up with something like that?

**Comment by heikkilv** [comment]
In my use case I receive the data as one day patches which contain multiple measurements for the attributes. I have used handwritten adapter program to send the data to Orion. Because I communicate directly with Orion I can of course use any data model I want (including any meta attributes). Since in my case the actual timestamps for the measurements where not the same as the sending time to Orion (<code>dateModified</code>), I used the [data model guidelines](https://fiware-datamodels.readthedocs.io/en/latest/guidelines/index.html) as a guide to how to store the attribute specific timestamps (i.e. <code>timestamp</code>) and found out that this was not supported by QuantumLeap.

The <code>myTime</code> example in the original post was just an example that someone can also use other time attributes (I have seen <code>dateMeasured</code> as well) than <code>TimeInstant</code> or <code>timestamp</code> and there should be some support for them as well.
